File: PythonSolution/src/Lv2Sol49993.py
#  Created By DevHyeon on 2020.11.25
#  Question Title : 스킬트리
#  Algorithm Type : Implementation
#  Level : 2
#  Link : https://programmers.co.kr/learn/courses/30/lessons/49993
def solution(skill, skill_trees):
    answer = 0

    for i in skill_trees:
        bool = True
        type = 0
        ind = 0
        # print(i)
        for j in skill:
            try:
                # print(i.index(j))
                if (i.index(j) >= ind and type == 0):
                    ind = i.index(j)


                else:
                    bool = False
            except ValueError as e:
                type = 1
                # print(e)

        # print(bool)
        if (bool == True):
            answer = answer + 1
            # print('=',answer)
    return answer

File: JavaSolution/src/programmers/Lv2Sol49993.java
package programmers;

public class Lv2Sol49993 {
    public int solution(String skill, String[] skill_trees) {
        int answer = 0;

        System.out.println("==="+skill+"===");

        for(String user_skill : skill_trees) {
            int startSkill = 0; //유저가 사전에 배워야 하는 스킬의 위치
            boolean isUse = true;
            for (int i=0; i<user_skill.length(); i++) {
                int tempSkill = skill.indexOf(user_skill.charAt(i)); //유저가 현재 배우려는 스킬
                //필수 스킬트리가 아니면 가능
                if (tempSkill == -1) {
                    continue;
                }
                //필수 스킬트리면 사전에 배워야하는 스킬과 같거나 작은지 확인
                else {
                    if (tempSkill <= startSkill) {
                        startSkill++;
                    }
                    else {
                        isUse = false;
                        break;
                    }
                }
            }
            if (isUse) {
                answer++;
            }
        }

        System.out.println(answer);
        return answer;
    }
}
